File: python/347. Top K Frequent Elements.py
from Queue import PriorityQueue
class Comparable(object):
    def __init__(self, fre, val):
        self.fre = fre
        self.val = val
    
    def __cmp__(self, other):
        if self.fre == other.fre:
            return 0
        elif self.fre > other.fre:
            return 1
        else:
            return -1

class Solution(object):
    def topKFrequent(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        mp = {}
        for num in nums:
            fre = mp.setdefault(num, 0)
            mp[num] = fre+1
        
        res = []
        que = PriorityQueue()
        for num, fre in mp.iteritems():
            que.put(Comparable(fre, num))
            if que.qsize() > k:
                que.get()

        while que.qsize():
            res.append(que.get().val)

        res.reverse()
        return res
        

if __name__ == '__main__':
    nums = [1,1,1,2,2,3]
    k = 2
    su = Solution()
    print su.topKFrequent(nums, k)

File: c++/347. Top K Frequent Elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> um;
        for (auto &i : nums)
            um[i]++;
        
        vector<int> res;
        priority_queue<pair<int, int>> pq;
        for (auto it = um.begin(); it != um.end(); it++) {
            pq.push(make_pair(it->second, it->first));
            if (pq.size() > um.size() - k) {
                res.push_back(pq.top().second);
                pq.pop();
            }
        }
        
        return res;
    }
};